spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java

```diff
@@ -159,6 +161,8 @@
 
 	private Set<PosixFilePermission> permissions;
 
+	private BiConsumer<File, Message<?>> newFileCallback;
```

Comment:
Perhaps it should be a `BiFunction<File, Message<?>, File>` with the callback returning the file we will actually write to.

Reply:
I am not sure I am following. This PR was supposed to handle writing some content into a newly created file. Why/when would we return a different file from the function? Thanks for the clarification.

Reply:
Let's stick with `BiConsumer` - no need to add complexity.

Reply:
I don't know what scenario may do such a mutation outside of the `FileWritingMessageHandler`, but let it is such a `BiFunction`!
The target writing is going to happen on the resulting file.
On the other hand we are going to to get such a mutated file only once: when a new file is created. All other subsequent appends are going to happen in the original file, evaluated by the `FileWritingMessageHandler`.
So, may be it is not so good idea to let end-user to mutate target file unconditionally.

Reply:
I was just generalizing - someone might want to change the file location for certain messages (not just write a header). But they can do that with directory expression and filename generator.